### PR TITLE
[nuvoton] Upgrade LVGL to v81

### PR DESCRIPTION
**Pull request by wosayttn**
## 拉取/合并请求描述：(PR description)

[
1. Support LVGLv81 RTT music demo on N9H30 and NUC980 platform.
- You also can find LVGLv81 with GE2D supporting implementation in https://github.com/wosayttn/lvgl/tree/release/v8.1 .

2. Update Drivers
]

以下的内容不应该在提交PR时的message修改，修改下述message，PR会被直接关闭。请在提交PR后，浏览器查看PR并对以下检查项逐项check，没问题后逐条在页面上打钩。
The following content must not be changed in the submitted PR message. Otherwise, the PR will be closed immediately. After submitted PR, please use a web browser to visit PR, and check items one by one, and ticked them if no problem.

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [ ] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [x] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [x] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [x] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other styles
- [x] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [x] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [x] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [x] 本拉取/合并请求代码是高质量的 Code in this PR is of high quality
- [x] 本拉取/合并使用[formatting](https://github.com/mysterywolf/formatting)等源码格式化工具确保格式符合[RT-Thread代码规范](../documentation/coding_style_cn.md) This PR complies with [RT-Thread code specification](../documentation/coding_style_en.txt)

### CODE REVIEW

File: bsp/nuvoton/nk-980iot/applications/lvgl/port/lv_port_disp.c [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        LOG_E("error!");
+        return;
+    }
+
+    lv_disp_drv_init(&disp_drv); /*Basic initialization*/
+
+    /*Set the resolution of the display*/
+    disp_drv.hor_res = info.width;
+    disp_drv.ver_res = info.height;
+    //disp_drv.full_refresh = 1;
```
Comment (mysterywolf) [inline bsp/nuvoton/nk-980iot/applications/lvgl/port/lv_port_disp.c:92]: 请把这个注释删除把 full_refresh会强制刷新全屏，没有什么用的参数

File: bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_port_disp.c [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        LOG_E("error!");
+        return;
+    }
+
+    lv_disp_drv_init(&disp_drv); /*Basic initialization*/
+
+    /*Set the resolution of the display*/
+    disp_drv.hor_res = info.width;
+    disp_drv.ver_res = info.height;
+    disp_drv.full_refresh = 1;
```
Comment (mysterywolf) [inline bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_port_disp.c:117]: 这个参数请删除

File: bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_port_indev.c [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+static rt_int16_t last_y = 0;
+
+static void input_read(lv_indev_drv_t *indev_drv, lv_indev_data_t *data)
+{
+    data->point.x = last_x;
+    data->point.y = last_y;
+    data->state = last_state;
+}
+
+void littlevgl2rtt_send_input_event(rt_int16_t x, rt_int16_t y, rt_uint8_t state)
```
Comment (mysterywolf) [inline bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_port_indev.c:27]: 请更改一下这个的函数的命名，littlevgl2rtt_send_input_event这个命名是错误的
please rename this function

File: bsp/nuvoton/nk-980iot/applications/lvgl/port/lv_port_indev.c [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+static rt_int16_t last_y = 0;
+
+static void input_read(lv_indev_drv_t *indev_drv, lv_indev_data_t *data)
+{
+    data->point.x = last_x;
+    data->point.y = last_y;
+    data->state = last_state;
+}
+
+void littlevgl2rtt_send_input_event(rt_int16_t x, rt_int16_t y, rt_uint8_t state)
```
Comment (mysterywolf) [inline bsp/nuvoton/nk-980iot/applications/lvgl/port/lv_port_indev.c:26]: please rename this function

File: bsp/nuvoton/nk-980iot/applications/lvgl/port/lv_port_disp.c [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    RT_ASSERT(info.bits_per_pixel == 8 || info.bits_per_pixel == 16 ||
+              info.bits_per_pixel == 24 || info.bits_per_pixel == 32);
+
+    buf_1 = (void *)info.framebuffer;
+    rt_kprintf("lv buf_1=%08x\n", buf_1);
+
+    /*Initialize `disp_buf` with the buffer(s).*/
+    lv_disp_draw_buf_init(&disp_buf, buf_1, NULL, info.width * info.height);
```
Comment (mysterywolf) [inline bsp/nuvoton/nk-980iot/applications/lvgl/port/lv_port_disp.c:78]: 如果内存空间够用，应该使用双buffer , buf_1 buf_2来提高fps

File: bsp/nuvoton/nk-980iot/applications/lvgl/port/lv_demo.c [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Change Logs:
+ * Date           Author        Notes
+ * 2021-10-17     Meco Man      First version
+ */
+#include <rtthread.h>
+#include <lvgl.h>
+#include <lv_port_indev.h>
```
Comment (mysterywolf) [inline bsp/nuvoton/nk-980iot/applications/lvgl/port/lv_demo.c:12]: 这个头文件没有用，请移除

File: bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_demo.c [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Change Logs:
+ * Date           Author        Notes
+ * 2021-10-17     Meco Man      First version
+ */
+#include <rtthread.h>
+#include <lvgl.h>
+#include <lv_port_indev.h>
```
Comment (mysterywolf) [inline bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_demo.c:12]: 这个头文件请移除

File: bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_demo.c [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    if (tid == RT_NULL)
+    {
+        LOG_E("Fail to create 'LVGL' thread");
+    }
+    rt_thread_startup(tid);
+
+    return 0;
+}
+INIT_APP_EXPORT(lvgl_demo);
+//MSH_CMD_EXPORT(lvgl_demo, start lvgl music demo);
```
Comment (mysterywolf) [inline bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_demo.c:51]: 请删除无用的注释

**Review by mysterywolf — CHANGES_REQUESTED** [review]

**Comment by wosayttn** [comment]
Thanks these comments. For full_refresh flag, I will verify the disabling before pushing next commit.

**Comment by mysterywolf** [comment]
我觉得你放在application里要比我放在board下面要更好，我也提pr改到application目录下！😄

File: bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_demo.c [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    lv_demo_music();
+
+    while (1)
+    {
+        lv_task_handler();
+        rt_thread_mdelay(1);
+    }
+}
+
+static int lvgl_demo(void)
```
Comment (mysterywolf) [inline bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_demo.c:37]: lvgl_demo_init

File: bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_demo.c [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    tid = rt_thread_create("LVGL", lvgl_thread, RT_NULL, LV_THREAD_STACK_SIZE, LV_THREAD_PRIO, 10);
+    if (tid == RT_NULL)
+    {
+        LOG_E("Fail to create 'LVGL' thread");
+    }
+    rt_thread_startup(tid);
+
+    return 0;
+}
+INIT_APP_EXPORT(lvgl_demo);
```
Comment (mysterywolf) [inline bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_demo.c:50]: lvgl_demo_init

**Review by mysterywolf — CHANGES_REQUESTED** [review]

**Comment by mysterywolf** [comment]
application/lvgl不必加port文件夹，直接application/lvgl即可

File: bsp/nuvoton/nk-980iot/applications/lvgl/port/lv_port_disp.c (hunk 1 above)
Comment (wosayttn) [reply bsp/nuvoton/nk-980iot/applications/lvgl/port/lv_port_disp.c]: Removed

File: bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_port_disp.c (hunk 2 above)
Comment (wosayttn) [reply bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_port_disp.c:115]: The flag still keep in my testing. 
image

File: bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_port_indev.c (hunk 3 above)
Comment (wosayttn) [reply bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_port_indev.c]: Renamed the callback function name to nu_touch_inputevent_cb.

File: bsp/nuvoton/nk-980iot/applications/lvgl/port/lv_port_indev.c (hunk 4 above)
Comment (wosayttn) [reply bsp/nuvoton/nk-980iot/applications/lvgl/port/lv_port_indev.c]: Renamed the callback function name to nu_touch_inputevent_cb.

File: bsp/nuvoton/nk-980iot/applications/lvgl/port/lv_demo.c (hunk 6 above)
Comment (wosayttn) [reply bsp/nuvoton/nk-980iot/applications/lvgl/port/lv_demo.c]: Removed

File: bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_demo.c (hunk 7 above)
Comment (wosayttn) [reply bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_demo.c]: Removed

File: bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_demo.c (hunk 8 above)
Comment (wosayttn) [reply bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_demo.c]: Removed

File: bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_demo.c (hunk 9 above)
Comment (wosayttn) [reply bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_demo.c]: Append

File: bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_demo.c (hunk 10 above)
Comment (wosayttn) [reply bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_demo.c]: Append

File: bsp/nuvoton/nk-980iot/applications/lvgl/port/lv_port_disp.c (hunk 5 above)
Comment (wosayttn) [reply bsp/nuvoton/nk-980iot/applications/lvgl/port/lv_port_disp.c]: Yes, it improves rendering FPS using two buffers.

File: bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_port_disp.c (hunk 2 above)
Comment (wosayttn) [reply bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_port_disp.c:115]: We take care tearing issue on our platform is with internal LCD controller. So, we still need to keep the flag.

Comment (mysterywolf) [reply bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_port_disp.c:115]: 请留意一下这里，LVGL作者的回复：
image
https://forum.lvgl.io/t/v8-display-driver-double-buffer-low-fps-high-cpu/6901/5
这就是为什么我对这个参数很敏感的原因，根据作者的回答，这个参数会强制刷新整个屏幕。

**Comment by mysterywolf** [comment]
处理完毕了请告诉我

File: bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_port_disp.c (hunk 2 above)
Comment (wosayttn) [reply bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_port_disp.c]: > 处理完毕了请告诉我

Done. Please revise the LVGLv81 porting on Nuvoton platform again.

File: bsp/nuvoton/nk-n9h30/applications/lvgl/lv_conf.h [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+#define LV_HOR_RES_MAX          (800)
+#define LV_VER_RES_MAX          (480)
+
+
+#define LV_USE_DEMO_RTT_MUSIC           1
+#define LV_USE_DEMO_MUSIC               1
+#define LV_FONT_MONTSERRAT_12           1
+#define LV_FONT_MONTSERRAT_16           1
+#define LV_DEMO_MUSIC_AUTO_PLAY         1
+#define LV_DEMO_RTT_MUSIC_AUTO_PLAY     1
```
Comment (mysterywolf) [inline bsp/nuvoton/nk-n9h30/applications/lvgl/lv_conf.h:28]: 这个位置有问题 用RTT定制版本的Demo，不必开启原始demo，参见：https://github.com/RT-Thread/rt-thread/blob/master/bsp/qemu-vexpress-a9/applications/lvgl/lv_conf.h#L17-L25
https://github.com/RT-Thread/rt-thread/blob/master/bsp/qemu-vexpress-a9/drivers/Kconfig#L23

File: bsp/nuvoton/nk-980iot/applications/lvgl/lv_conf.h [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+#define LV_HOR_RES_MAX          (320)
+#define LV_VER_RES_MAX          (240)
+
+
+#define LV_USE_DEMO_RTT_MUSIC           1
+#define LV_USE_DEMO_MUSIC               1
+#define LV_FONT_MONTSERRAT_12           1
+#define LV_FONT_MONTSERRAT_16           1
+#define LV_DEMO_MUSIC_AUTO_PLAY         1
+#define LV_DEMO_RTT_MUSIC_AUTO_PLAY     1
```
Comment (mysterywolf) [inline bsp/nuvoton/nk-980iot/applications/lvgl/lv_conf.h:27]: 此处也是

**Review by mysterywolf — CHANGES_REQUESTED** [review]

File: bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_port_disp.c (hunk 2 above)
Comment (wosayttn) [reply bsp/nuvoton/nk-n9h30/applications/lvgl/port/lv_port_disp.c]: We also implemented direct_mode, full_refresh and two screen-sized buffers in the N9H30 porting. Let user has more choices.

File: bsp/nuvoton/nk-n9h30/applications/lvgl/lv_conf.h (hunk 11 above)
Comment (wosayttn) [reply bsp/nuvoton/nk-n9h30/applications/lvgl/lv_conf.h]: Done

File: bsp/nuvoton/nk-980iot/applications/lvgl/lv_conf.h (hunk 12 above)
Comment (wosayttn) [reply bsp/nuvoton/nk-980iot/applications/lvgl/lv_conf.h]: Done

**Comment by mysterywolf** [comment]
https://github.com/RT-Thread/rt-thread/blob/master/bsp/stm32/stm32l475-atk-pandora/board/Kconfig#L63-L67 需要给用户提供一个一键化配置LVGL的选项，在onboard配置项里，名字叫做“Enable LVGL for LCD”

**Comment by wosayttn** [comment]
> https://github.com/RT-Thread/rt-thread/blob/master/bsp/stm32/stm32l475-atk-pandora/board/Kconfig#L63-L67 需要给用户提供一个一键化配置LVGL的选项，在Hardware Drivers Config->Onboard Peripheral Drivers配置项里，名字叫做“Enable LVGL for LCD”

Hi,

Both LVGLv81 core and rtt_music_demo packages are enabled in default configuration file - .config.
So, the rtt_music_demo demonstrates on the board by default.

https://github.com/RT-Thread/rt-thread/pull/5399#issuecomment-1001889052

**Comment by mysterywolf** [comment]
@wosayttn 这么做也可以，但是没有明确的依赖关系，用户去除掉这个功能很麻烦。我们现在的策略一般是默认这种附加功能不选定，如果用户需要，再选定，选定的时候Kconfig会自动选定依赖项。不要默认就把一堆的东西都开开起来，尤其是软件包什么的。

**Review by mysterywolf — APPROVED** [review]

**Comment by mysterywolf** [comment]
下个pr再说这个问题吧

**Comment by wosayttn** [comment]
Thanks reviewer and merging.

**Pull request merged**